erasure-coding/src/lib.rs

```diff
@@ -0,0 +1,409 @@
+	let mut trie_storage: MemoryDB<Blake2Hasher> = MemoryDB::default();
+	for node in branch_nodes.iter() {
+		(&mut trie_storage as &mut trie::HashDB<_>).insert(node.as_slice());
+	}
+
+	let trie = TrieDB::new(&trie_storage, &root).map_err(|_| Error::InvalidBranchProof)?;
+	let res = (index as u32).using_encoded(|key|
+		trie.get_with(key, |raw_hash: &[u8]| H256::decode(&mut &raw_hash[..]))
+	);
+
+	match res {
+		Ok(Some(Some(hash))) => Ok(hash),
+		Ok(Some(None)) => Err(Error::InvalidBranchProof), // hash failed to decode
+		Ok(None) => Err(Error::BranchOutOfBounds),
+		Err(_) => Err(Error::InvalidBranchProof),
+	}
+}
+
+// input for `parity_codec` which draws data from the data shards
+struct ShardInput<'a, I> {
+	shards: I,
+	cur_shard: Option<(&'a [u8], usize)>,
+}
+
+impl<'a, I: Iterator<Item=&'a [u8]>> codec::Input for ShardInput<'a, I> {
+	fn read(&mut self, into: &mut [u8]) -> usize {
+		let mut read_bytes = 0;
+
+		loop {
+			if read_bytes == into.len() { break }
+
+			let cur_shard = self.cur_shard.take().or_else(|| self.shards.next().map(|s| (s, 0)));
+			let (active_shard, mut in_shard) = match cur_shard {
+				Some((s, i)) => (s, i),
+				None => break,
+			};
+
+			if in_shard >= active_shard.len() {
+				continue;
+			}
+
+			let remaining_len_out = into.len() - read_bytes;
+			let remaining_len_shard = active_shard.len() - in_shard;
+
+			let write_len = std::cmp::min(remaining_len_out, remaining_len_shard);
+			into[read_bytes..][..write_len]
+				.copy_from_slice(&active_shard[in_shard..][..write_len]);
+
+			in_shard += write_len;
+			read_bytes += write_len;
+			self.cur_shard = Some((active_shard, in_shard))
+		}
+
+		read_bytes
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+
+	#[test]
+	fn field_order_is_right_size() {
+		assert_eq!(MAX_VALIDATORS, 65536);
+	}
+
+    #[test]
+	fn round_trip_block_data() {
+		let block_data = BlockData((0..255).collect());
+		let ex = Extrinsic { outgoing_messages: Vec::new() };
+		let chunks = obtain_chunks(
+			10,
+			&block_data,
+			&ex,
+		).unwrap();
+
+		assert_eq!(chunks.len(), 10);
+
+		// any 4 chunks should work.
+		let reconstructed = reconstruct(
+			10,
+			[
+				(&*chunks[1], 1),
+				(&*chunks[4], 4),
+				(&*chunks[6], 6),
+				(&*chunks[9], 9),
+			].iter().cloned(),
+		).unwrap();
+
+		assert_eq!(reconstructed, (block_data, ex));
+	}
+
+	#[test]
+	fn construct_valid_branches() {
+		let block_data = BlockData(vec![2; 256]);
+		let chunks = obtain_chunks(
+			10,
+			&block_data,
+			&Extrinsic { outgoing_messages: Vec::new() },
+		).unwrap();
+		let chunks: Vec<_> = chunks.iter().map(|c| &c[..]).collect();
+
+		assert_eq!(chunks.len(), 10);
+
+		let branches = branches(chunks.clone());
+		let root = branches.root();
+
+		let proofs: Vec<_> = branches.map(|(proof, _)| proof).collect();
+
+		assert_eq!(proofs.len(), 10);
+
+		for (i, proof) in proofs.into_iter().enumerate() {
```

Comment:
I'm not sure, I'm getting this correctly. I think each Merkle proof should contains the hash value of the siblings of all nodes in your branch. So you can compute your hash way up to the root. Am I missing something?

Reply:
we are just including all the nodes on the path -- that's how you generally do merkle proofs for 16-radix (which we are using here)

Reply:
I verified with @burdges, you need the co-path not the path. On a binary tree it is all the other children of the nodes on the path.

Reply:
@burdges why do we need the co-path? the path seems to work (although can be optimized by omitting the hash of the child we are traversing to from branches)

Reply:
(have chatted with @burdges in Riot and confirmed that it is correct as-is _for the patricia trie_)